### PR TITLE
Add Start/Loop/End sounds for weapons

### PR DESCRIPTION
**Pull request by Baezon**
Rather than just repeated launch sounds, can make high rate rate weapons sound better. I really wanted to make this work for turrets but damn it's a pain in the ass, so just primaries/secondaries will have to do for now.

### CODE REVIEW

File: code/ship/ship.cpp [hunk 1]
```diff
@@ -10118,6 +10120,66 @@ static void ship_radar_process( object * obj, ship * shipp, ship_info * sip )
...
+
+	for (int i = 0; i < swp->num_primary_banks; i++) {
+		if (swp->primary_bank_weapons[i] < 0)
+			continue;
+
+		weapon_info* wip = &Weapon_info[swp->primary_bank_weapons[i]];
+
+		bool primaries_locked = shipp->flags[Ship_Flags::Primaries_locked];
+		bool selected = swp->current_primary_bank == i || (shipp->flags[Ship_Flags::Primary_linked] && !wip->wi_flags[Weapon::Info_Flags::Nolink]);
+		bool has_resources = shipp->weapon_energy >= wip->energy_consumed && (!wip->wi_flags[Weapon::Info_Flags::Ballistic] || swp->primary_bank_ammo[i] >= 0);
```
Comment (JohnAFernandez) [inline code/ship/ship.cpp:10140]: Is swp->primary_bank_ammo[i] >= 0 intended or should it be >0?

File: code/ship/ship.cpp [hunk 2]
```diff
@@ -10118,6 +10120,66 @@ static void ship_radar_process( object * obj, ship * shipp, ship_info * sip )
...
+
+			vec3d pos = model_get(Ship_info[shipp->ship_info_index].model_num)->view_positions[0].pnt;
+
+			if (wip->loop_firing_snd.isValid())
+				swp->firing_loop_sounds[i] = obj_snd_assign(shipp->objnum, wip->loop_firing_snd, &pos, OS_PLAY_ON_PLAYER);
+			else
+				swp->firing_loop_sounds[i] = -2;
+		} 
+
+		if (swp->firing_loop_sounds[i] != -1 && (!trigger_down || primaries_locked || !selected || !has_resources || !burst_only_allowed || dying)) {
```
[the remark this answers is not in the document]
Comment (Baezon) [reply code/ship/ship.cpp:10164]: Yes, as you said, its for playing the ending sound, and possibly ending the loop sound. All 3 sounds are optional, and can be combined in any order, having an end or start sound does not imply you've given or want a loop sound.

File: code/ship/ship.cpp (hunk 1 above)
Comment (Baezon) [reply code/ship/ship.cpp:10140]: Whoops, yes.

File: code/ship/ship_flags.h [hunk 3]
```diff
@@ -8,8 +8,9 @@ namespace Ship {
 	FLAG_LIST(Weapon_Flags){
 		Beam_Free = 0,	// if this is a beam weapon, its free to fire
 		Turret_Lock,	// is this turret is free to fire or locked
-		Tagged_Only,	// only fire if target is tagged
-		Trigger_Lock,	// // indicates that the trigger is held down
+		Tagged_Only,	// only fire if target is tagged		
+		Secondary_trigger_down,	// indicates that the trigger is held down, used for multilock hold-to-lock type missiles
+		Primary_trigger_down,	// indicates that the trigger is held down
```
[the remark this answers is not in the document]
Comment (Baezon) [reply code/ship/ship_flags.h:13]: I mean, it makes logical sense to me. The trigger state is kinda part of the weapons state. Eventually, I want to make this work for turrets too, and they have their own `ship_weapon` as well, with their own trigger states.

**Review by JohnAFernandez — APPROVED** [review]

**Review by Goober5000 — COMMENTED** [review]
Does this replace or supplement the `was_firing_last_frame` checks?  That was introduced as part of Bobboau's fighter beams - one of the earliest FSO features, and one which required contributions from several coders to get working - but it looks like the "trigger down" flag is a better way to handle it.

**Comment by Baezon** [comment]
Right now the loop_sounds tracking is entirely orthogonal, which is unideal, but `was_firing_last_frame` tracks slightly different conditions, and I was hesitant to change it's behavior to make the new behavior work as well. `Primary_trigger_down` at least, isn't meant to be bank specific or anything, just the generalized 'is this entity trying to fire'.

**Comment by wookieejedi** [comment]
Will plan to merge this later today unless someone beats me to it

**Pull request merged**